crates/goose/src/providers/embedded.rs

```diff
@@ -0,0 +1,631 @@
+            let stages = vec![
+                SamplerStage::Temperature(temperature),
+                SamplerStage::TopK(40),
+                SamplerStage::TopP(0.9),
+            ];
+            let sampler = StandardSampler::new_softmax(stages, 1);
+            
+            // Start completion
+            let mut completion_handle = session
+                .start_completing_with(sampler, max_tokens as usize)
+                .map_err(|e| anyhow::anyhow!("Failed to start completion: {}", e))?;
+            
+            let mut generated_text = String::new();
+            let mut token_count = 0;
+            
+            // Generate tokens
+            while let Some(token) = completion_handle.next_token() {
+                let token_string = session.model().token_to_piece(token);
+                generated_text.push_str(&token_string);
+                token_count += 1;
+                
+                if token_count >= max_tokens as usize {
+                    break;
+                }
+                
+                // Check for stop sequences
+                let mut hit_stop = false;
+                for stop_seq in &stop_sequences {
+                    if generated_text.contains(stop_seq) {
+                        hit_stop = true;
+                        break;
+                    }
+                }
+                
+                if hit_stop {
+                    break;
+                }
+            }
+            
+            // Clean up stop sequences from the end
+            for stop_seq in &stop_sequences {
+                if let Some(pos) = generated_text.find(stop_seq) {
+                    generated_text.truncate(pos);
+                    break;
+                }
+            }
+            
+            Ok(generated_text.trim().to_string())
+        })
+        .await
+        .map_err(|e| ProviderError::ExecutionError(format!("Task join error: {}", e)))?
+        .map_err(|e| ProviderError::ExecutionError(e.to_string()))?;
+        
+        Ok(result)
+    }
+    
+
+}
+
+#[async_trait]
+impl Provider for EmbeddedProvider {
+    fn metadata() -> ProviderMetadata {
+        ProviderMetadata::new(
+            "embedded",
+            "Embedded",
+            "Run local language models with optional tool execution",
+            EMBEDDED_DEFAULT_MODEL,
+            EMBEDDED_KNOWN_MODELS.to_vec(),
+            "",
+            vec![],
+        )
+    }
+    
+    fn get_model_config(&self) -> ModelConfig {
+        self.model.clone()
+    }
+    
+    #[tracing::instrument(
+        skip(self, model_config, system, messages),
+        fields(model_config, input, output, input_tokens, output_tokens, total_tokens)
+    )]
+    async fn complete_with_model(
+        &self,
+        model_config: &ModelConfig,
+        system: &str, // Ignored - we use g3's system prompt for embedded
+        messages: &[Message],
+        _tools: &[Tool], // Ignored - embedded provider doesn't use tool definitions
```

Comment:
going forward, is this an option? it seems through ollama we'd get somewhere

Reply:
(I think have to add in BNF grammars and stuff if we wanted to do out of the box what ollama et al do? valuable?)

Reply:
ollama seems to cope, but others struggle due to the system prompt+tools now, so if we want ot to run lean, really need to strip them out somehow